splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java

```diff
@@ -0,0 +1,282 @@
+import org.supercsv.prefs.CsvPreference;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.util.List;
+
+/**
+ * A Tokenizer which keeps track of which columns were quoted and which were not.
+ * This allows us to make a distinction between columns with value {@code Foo}
+ * versus columns with values {@code "Foo"}.
+ *
+ * @author Scott Fines
+ *         Date: 9/29/16
+ */
+public class QuoteTrackingTokenizer extends AbstractTokenizer{
+
+    private static final char NEWLINE='\n';
+
+    private static final char SPACE=' ';
+
+    private final StringBuilder currentColumn=new StringBuilder();
+
+    /* the raw, untokenized CSV row (may span multiple lines) */
+    private final StringBuilder currentRow=new StringBuilder();
+
+    private final int quoteChar;
+
+    private final int delimeterChar;
+
+    private final boolean surroundingSpacesNeedQuotes;
+
+    private final boolean ignoreEmptyLines;
+
+    private final CommentMatcher commentMatcher;
+
+    private final int maxLinesPerRow;
+
+    /**
+     * Enumeration of tokenizer states. QUOTE_MODE is activated between quotes.
+     */
+    private enum TokenizerState{
+        NORMAL,QUOTE_MODE;
+    }
+
+    /**
+     * Constructs a new <tt>Tokenizer</tt>, which reads the CSV file, line by line.
+     *
+     * @param reader      the reader
+     * @param preferences the CSV preferences
+     * @throws NullPointerException if reader or preferences is null
+     */
+    public QuoteTrackingTokenizer(final Reader reader,final CsvPreference preferences){
+        super(reader,preferences);
+        this.quoteChar=preferences.getQuoteChar();
+        this.delimeterChar=preferences.getDelimiterChar();
+        this.surroundingSpacesNeedQuotes=preferences.isSurroundingSpacesNeedQuotes();
+        this.ignoreEmptyLines=preferences.isIgnoreEmptyLines();
+        this.commentMatcher=preferences.getCommentMatcher();
+        this.maxLinesPerRow=preferences.getMaxLinesPerRow();
+    }
+
+    @Override
+    public boolean readColumns(final List<String> columns) throws IOException{
+        return readColumns(columns,null);
+    }
+
+    public boolean readColumns(final List<String> columns,final BooleanList quotedColumns) throws IOException{
+        if(columns==null){
+            throw new NullPointerException("columns should not be null");
+        }
+        boolean recordQuotes = quotedColumns!=null;
+
+        // clear the reusable List and StringBuilders
+        columns.clear();
+        if(recordQuotes) quotedColumns.clear();
+        currentColumn.setLength(0);
+        currentRow.setLength(0);
+
+        // read a line (ignoring empty lines/comments if necessary)
+        String line;
+        do{
+            line=readLine();
+            if(line==null){
+                return false; // EOF
+            }
+        }
+        while(ignoreEmptyLines && line.length()==0 || (commentMatcher!=null && commentMatcher.isComment(line)));
```

Comment:
is it possible that line may contain other white spaces, or it needs trim before length check?